wallet-new/src/Cardano/Wallet/Kernel/BListener.hs

```diff
@@ -77,15 +79,33 @@ switchToFork pw@PassiveWallet{..} n bs = do
     k <- Node.getSecurityParameter _walletNode
     blocksAndMeta <- mapM (prefilterBlock' pw) bs
     let (blockssByAccount, metas) = unzip blocksAndMeta
-    res <- update' _wallets $ SwitchToFork k n blockssByAccount
```

Comment:
I think we have to be more careful with atomicity here. Instead I suggest:

1. We do _not_ try to hold the `walletRestorationTask` `MVar` while we do the switch to fork, because this does not guarantee that a new restoration somewhere won't modify the DB, so it doesn't really give us very much.
2. Instead we call on all current restoration tasks a new function `prepareForRestart`, which will give us a wallet with the initial account and initial UTxO.
3. When we call `switchToFork`, we pass in as argument a map from `WalletId` to the initial wallets we got from `prepareForRestart`. Whenever `switchToFork` encounters an account in incomplete state, it replaces that wallet with the initial wallet. If we are missing an initial wallet, it must mean another restoration got started; in this case, `switchToFork` will return an error and we basically just go again, filling in the blanks for the missing wallets.
4. Restoration tasks that try to apply a historical block to a checkpoint and fail ("not successor") will just kill themselves (this would be the result of `SwitchToFork` switching out the wallet).
5. After switch to work we call `restartRestoration` all all of these workers.